AEP/aep-00004.md

```diff
@@ -0,0 +1,163 @@
+
+Finally, the current MongoDB module in our infrastructure's code is defined kind of manually with multiple resources being defined, which is not  necessarily the best choice as deploying a database, especially a sharded one, is a complex task. 
+
+# Rationale
+Sharding the database would alleviate the tension on all the nodes. For the primary nodes, the write operations would be distributed among other shards' primary node, and for the secondary nodes there can be some situations where only the right shard is queried (it will be explicited below in this AEP) and as mentionned above, the read operations are faster on a sharded cluster since the data is distributed among multiple instances.
+
+This sharded (or not) MongoDB deployment would be defined in our infrastructure code as a Helm release resource (called by Terraform). A Helm release is an instance of a Helm chart. Helm charts are templates that allow to deploy resources within a Kubernetes cluster in configurable way.
+Using Helm charts conveys the complexity of defining a deployment towards the authors of a Helm chart who had more time to think about how MongoDB should be deployed within a Kubernetes cluster.
+
+Instead of  manually defining a MongoDB deployment with Terraform modules as it is currently done with ArmoniK's infrastructure, Helm charts do it for us. All we have to do is to define the right configuration for the MongoDB deployment.
+
+During our reflexion, some other alternatives to responding this issue were considered :
+- First, adding some resources and modifying some others in our Terraform module. But according to the KISS and "don't reinvent the wheel" principles, we found this solution unrelevant as others developers are able to provide us out-of-the-box MongoDB deployments, which is often a better way of building softwares as the complexity is shifted to the Helm chart definiton.
+- Using Kubernetes operator for MongoDB, but as far as we have explored this solution, we were not able to find an operator that supports sharding.
+
+To conclude, using a MongoDB Helm chart would allow ArmoniK to benefit from an out-of-the-box MongoDB deployment to separate the complexity of it from the sheer infrastructure code defined in Terraform, as well as benefit from the maintenance of the Helm chart used, again in the scope of dividing complexity.
+
+# Specification
+
+## Overview
+As ArmoniK's infrastructure is described via Terraform, instead of defining multiple Kubernetes and other resources as it is currently done, we would (almost) only define a `helm_release` resource provided by the Helm provider in Terraform. The "almost" is there because we still need to define two kubernetes secrets that store the MongoDB deployment's information and users' credentials to ensure retrocompatibility with the way ArmoniK currently retrieves these data.
```

Comment:
```suggestion
Until now ArmoniK's infrastructure was mainly described using Terraform files and bash scripts. Using Helm charts 
for ArmoniK's deployement on Kubernetes will reduce the maintenance and improve the reusability of our example infrastructure by other organisation.
```

Reply:
ArmoniK's example infrastructure will still be described using Terraform files ! That said, this paragraph should rather be about ArmoniK's table storage than about its whole infrastructure.